libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -517,6 +517,109 @@ string YulUtilFunctions::arrayLengthFunction(ArrayType const& _type)
 	});
 }
 
+std::string YulUtilFunctions::resizeDynamicArrayFunction(ArrayType const& _type)
+{
+	solAssert(_type.location() == DataLocation::Storage, "");
+	solAssert(_type.isDynamicallySized(), "");
+	if (!_type.isByteArray() && _type.baseType()->storageBytes() < 32)
+		solAssert(_type.baseType()->isValueType(), "Invalid storage size for non-value type.");
+
+	string functionName = "resize_array_" + _type.identifier();
+	return m_functionCollector->createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, newLen) {
+				let oldLen := <fetchLength>(array)
+				<?isByteArray>
+				// For a "long" byte array, store length as 2*length+1
+				newLen := add(1, mul(newLen, 2))
+				</isByteArray>
+
+				// Store new length
+				sstore(array, newLen)
+
+				// Size was reduced, clear end of array
+				if lt(newLen, oldLen) {
+					let oldSize := <convertToSize>(oldLen)
+					let newSize := <convertToSize>(newLen)
+					let arrayDataStart := <dataPosition>(array)
+					let deleteStart := add(arrayDataStart, newSize)
+					let deleteEnd := add(arrayDataStart, oldSize)
+					<clearStorageRange>(deleteStart, deleteEnd)
+				}
+			})")
+			("functionName", functionName)
+			("fetchLength", arrayLengthFunction(_type))
+			("isByteArray", _type.isByteArray())
+			("convertToSize", arrayConvertLengthToSize(_type))
+			("dataPosition", arrayDataAreaFunction(_type))
+			("clearStorageRange", clearStorageRangeFunction(*_type.baseType()))
+			.render();
+	});
+}
+
+string YulUtilFunctions::clearStorageRangeFunction(Type const& _type)
+{
+	string functionName = "clear_storage_range_" + _type.identifier();
+
+	return m_functionCollector->createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(start, end) {
+				for {} iszero(gt(start, end)) { start := add(start, <storageSize>) }
+				{
+					<updateStorageValue>(start, 0, <zeroValue>())
```

Comment:
This looks a bit fishy to me - is this supposed to work for types that are packed in storage? To me it looks like this will only clear offset 0 of each packed slot, right? And it'll do the whole ``sstore(..., shift_and_mask(sload(...), 0))`` thing - which you really only want to do once, in case ``start`` is not on a slot boundary and the rest you'd just want to clear with simple ``sstore(..., 0)``s? Or was the idea to let the optimizer handle all that (which will probably be extremely hard here)?

Reply:
Hmm once again, I only concentrated on 32 byte types (uint256 and such) with the plan to fix it once the first part works.
I'll change it to unimplementedAssert if storagesize is not 1 and use a simple `sstore()` here.